bee-test/src/rand/output/feature_block.rs

```diff
@@ -0,0 +1,86 @@
+pub fn rand_sender_feature_block() -> SenderFeatureBlock {
+    SenderFeatureBlock::new(rand_address())
+}
+
+/// Generates a random [`IssuerFeatureBlock`].
+pub fn rand_issuer_feature_block() -> IssuerFeatureBlock {
+    IssuerFeatureBlock::new(rand_address())
+}
+
+/// Generates a random [`DustDepositReturnFeatureBlock`].
+pub fn rand_dust_deposit_return_feature_block() -> DustDepositReturnFeatureBlock {
+    DustDepositReturnFeatureBlock::new(rand_number_range(DUST_DEPOSIT_MIN..IOTA_SUPPLY)).unwrap()
+}
+
+/// Generates a random [`TimelockMilestoneIndexFeatureBlock`].
+pub fn rand_timelock_milestone_index_feature_block() -> TimelockMilestoneIndexFeatureBlock {
+    TimelockMilestoneIndexFeatureBlock::new(rand_milestone_index())
+}
+
+/// Generates a random [`TimelockUnixFeatureBlock`].
+pub fn rand_timelock_unix_feature_block() -> TimelockUnixFeatureBlock {
+    TimelockUnixFeatureBlock::new(rand_number())
+}
+
+/// Generates a random [`ExpirationMilestoneIndexFeatureBlock`].
+pub fn rand_expiration_milestone_index_feature_block() -> ExpirationMilestoneIndexFeatureBlock {
+    ExpirationMilestoneIndexFeatureBlock::new(rand_milestone_index())
+}
+
+/// Generates a random [`ExpirationUnixFeatureBlock`].
+pub fn rand_expiration_unix_feature_block() -> ExpirationUnixFeatureBlock {
+    ExpirationUnixFeatureBlock::new(rand_number())
+}
+
+/// Generates a random [`MetadataFeatureBlock`].
+pub fn rand_metadata_feature_block() -> MetadataFeatureBlock {
+    let bytes = rand_bytes(rand_number_range(1..MetadataFeatureBlock::LENGTH_MAX));
+    MetadataFeatureBlock::new(&bytes).unwrap()
+}
+
+/// Generates a random [`IndexationFeatureBlock`].
+pub fn rand_indexation_feature_block() -> IndexationFeatureBlock {
+    let bytes = rand_bytes(rand_number_range(1..IndexationFeatureBlock::LENGTH_MAX));
+    IndexationFeatureBlock::new(&bytes).unwrap()
+}
+
+fn all_feature_blocks() -> Vec<FeatureBlock> {
+    vec![
+        FeatureBlock::Sender(rand_sender_feature_block()),
+        FeatureBlock::Issuer(rand_issuer_feature_block()),
+        FeatureBlock::DustDepositReturn(rand_dust_deposit_return_feature_block()),
+        FeatureBlock::TimelockMilestoneIndex(rand_timelock_milestone_index_feature_block()),
+        FeatureBlock::TimelockUnix(rand_timelock_unix_feature_block()),
+        FeatureBlock::ExpirationMilestoneIndex(rand_expiration_milestone_index_feature_block()),
+        FeatureBlock::ExpirationUnix(rand_expiration_unix_feature_block()),
+        FeatureBlock::Metadata(rand_metadata_feature_block()),
+        FeatureBlock::Indexation(rand_indexation_feature_block()),
+    ]
+}
+
+/// Generates a [`Vec`] of random [`FeatureBlock`]s given a set of allowed [`FeatureBlockFlags`].
+pub fn rand_allowed_feature_blocks(allowed_feature_blocks: FeatureBlockFlags) -> Vec<FeatureBlock> {
+    let mut all_feature_blocks = all_feature_blocks();
+    all_feature_blocks.retain(|feature_block| allowed_feature_blocks.contains(feature_block.flag()));
```

Comment:
I'm just thinking out loud but isn't there a way to do the opposite ? From the allowed flags generate the feature blocks ? That would mean matching I guess. I think the current way is a tiny bit error prone because we would have to remember to modify `all_feature_blocks` whenever we have new blocks, the other way we wouldn't have to.

Reply:
That's a very good point which is closely intertwined with our discussion on the `KIND` associated constant. It would be great if things like uniqueness and ordering of `KIND`s were checked at compile time. One way could be to have a central register (maybe in the form of a macro) that hands out the `KIND`s (in the form of trait implementations).

But until we have such a system, I think you're right we should turn it around.

Reply:
The problem is though, that I'm not aware if there is a way to iterate over all possible values in a bitflag (apart from trying all digits in the integer that represents the bitflag).

Reply:
I think we are basically trying to abstract over what an `enum` "is". So if we have
```rust
struct Bar;

struct Baz;

enum Foo {
    Bar(Bar),
    Baz(Baz),
}
```
we want to somehow assign `Bar` and `Baz` a `KIND` so we can distinguish them. But at the same time `Foo` should guarantee that both `KIND`s are different. This is basically what a [discriminant](https://doc.rust-lang.org/core/mem/struct.Discriminant.html) is but we want to control its value too.

An option would be to introduce two traits
```rust
trait Kinded {
    type Kind;
    // This could be a constant.
    fn kind(&self) -> Self::Kind;
}

trait KindOf<K: Kinded> {
     const KIND: K::Kinded;
}
```
`KindOf` would be implemented for `Bar` and `Baz` and `Kinded` could be derived for `Foo` to produce the following code:

```rust
impl Kinded for Foo {
    type Kind = u8;
    
    fn kind(&self) -> Self::Kind {
        const _: () = {
            if Bar::KIND == Baz::KIND {
                panic!("here be dragons");
            }
        };
        match self {
            Self::Bar(_) => Bar::KIND,
            Self::Baz(_) => Baz::KIND,
        }
    }
}

Reply:
I am not very fond of these traits tbh. It makes everything more complex. If the only thing we are trying to achieve with them is to be sure there are no duplicates I wouldn't worry too much about it. I really don't see how we can mess up that much. I was more interested in you having a look at the second part of the comment :p